File: ParkingLot/ParkingLot_Go/parkingLot/parkingLot.go
package parkingLot

import (
	"parkingLot/level"
	"parkingLot/vehicles"
)

var parkingLot *ParkingLot

func GetNewParkingLot() *ParkingLot {
	if parkingLot == nil {
		parkingLot = &ParkingLot{
			floors: 0,
			levels: make([]level.Level, 0),
		}
	}
	return parkingLot
}

type ParkingLot struct {
	floors int
	levels []level.Level
}

func (p *ParkingLot) AddLevel(spots int) {
	level := level.GetNewLevel(p.floors, spots)
	p.levels = append(p.levels, *level)
	p.floors++
}

func (p *ParkingLot) ParkVehicle(v vehicles.Vehicle) bool {
	for _, level := range p.levels {
		if level.ParkVehicle(v) {
			return true
		}
	}
	return false
}

func (p *ParkingLot) UnParkVehicle(v vehicles.Vehicle) bool {
	for _, level := range p.levels {
		if level.UnParkVehicle(v) {
			return true
		}
	}
	return false
}

func (p *ParkingLot) CheckAvailability() int {
	count := 0
	for _, level := range p.levels {
		count += level.CheckAvailability()
	}
	return count
}

File: ParkingLot/ParkingLot_Node/parkingLot.ts
import Level from "./level.ts";
import { Vehicle } from "./vehicles/vehicle.ts";

export default class ParkingLot {
  static _instance: ParkingLot;

  _floorCount: number;
  _levels: Level[];

  private constructor() {
    this._floorCount = 0;
    this._levels = [];
  }

  static get instance() {
    if (!this._instance) {
      this._instance = new ParkingLot();
    }
    return this._instance;
  }

  get floorCount() {
    return this._floorCount;
  }

  get levels() {
    return this._levels;
  }

  addLevel(spotsCount: number) {
    this._levels.push(new Level(this.floorCount, spotsCount));
  }

  parkVehicle(v: Vehicle) {
    const n = this.levels.length;
    for (let i = 0; i < n; i++) {
      const level = this.levels[i];
      if (level.parkVehicle(v)) {
        return true;
      }
    }
    return false;
  }

  unParkVehicle(v: Vehicle) {
    const n = this.levels.length;
    for (let i = 0; i < n; i++) {
      const level = this.levels[i];
      if (level.unParkVehicle(v)) {
        return true;
      }
    }
    return false;
  }

  checkAvailability() {
    let count = 0;
    const n = this.levels.length;
    for (let i = 0; i < n; i++) {
      const level = this.levels[i];
      count += level.checkAvailability();
    }
    return count;
  }
}
